src/pipeformer/internal/resolve.py

```diff
@@ -0,0 +1,199 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Helpers for resolving custom formatting."""
+from typing import Iterable, Union
+
+import attr
+from attr.validators import deep_mapping, instance_of
+from troposphere import Join
+
+from .structures import Input
+
+_INPUT_TAG = ["{INPUT:", "}"]
+_PRIMITIVE_TYPES = (int, float, complex, bool, type(None))
+_PrimitiveTypes = Union[int, float, complex, bool, None]
+__all__ = ("InputResolver",)
+
+
+def _tag_in_string(source: str, start: str, end: str) -> bool:
+    """Determine if a specific tag is in a string.
+
+    :param source: String to evaluate
+    :param start: String that marks the start of a tag
+    :param end: String that marks the end of a tag
+    :returns: Decision
+    """
+    if start not in source:
+        return False
+
+    if end not in source[source.index(start) + len(start) :]:
+        return False
+
+    return True
+
+
+def _value_to_triplet(source: str, start: str, end: str) -> Iterable[str]:
+    """Extract the first tag value from a string, splitting the source string into the parts before and after the tag.
+
+    :param source: String to process
+    :param start: String that marks the start of a tag
+    :param end: String that marks the end of a tag
+    :return: Split string values
+    """
+    prefix, _value = source.split(start, 1)
+
+    value, suffix = _value.split(end, 1)
+
+    return prefix, value, suffix
+
+
+@attr.s(cmp=False)
+class InputResolver:
+    """Wraps another structure and injects input references whenever a value is found that contains an input tag.
+
+    As strings are read from the contents of the wrapped structure,
+    they are expanded as necessary to CloudFormation dynamic references that will resolve the needed input values.
+
+    Along the way, the referenced inputs are collected and can later be found in ``required_inputs``.
+    This can be used to determine what inputs are required as parameters for a given CloudFormation template.
+
+    :param wrapped: Wrapped structure
+    :param inputs: Map of input names to :class:`Input` structures
+    :param required_inputs: Known required input (optional)
+    """
+
+    _wrapped = attr.ib()
+    _inputs = attr.ib(validator=deep_mapping(key_validator=instance_of(str), value_validator=instance_of(Input)))
+    required_inputs = attr.ib(default=attr.Factory(set))
+
+    @_wrapped.validator
```

Comment:
Minor: a short explanatory comment might make this easier to read for non-attrs-wizards, but it also does make sense with some squinting, so up to you.